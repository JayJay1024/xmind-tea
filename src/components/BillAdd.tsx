import React, { useState } from 'react';
import {
    Button,
    Select,
    DatePicker,
    InputNumber,
    message
} from 'antd';

import {
    TypeBill,
    TypeCategory,
} from '../types';
import '../styles.css';

type TypeProp = {
    dataCategory: TypeCategory[],
    onClick: (billNew: TypeBill) => void
};

const BillAdd: React.FunctionComponent<TypeProp> = (props): JSX.Element => {
    const [billTime, setBillDate]         = useState<Date>(new Date(0));  // 日期
    const [billCategory, setBillCategory] = useState<string>('');         // 类型
    const [billAmount, setBillAmount]     = useState<number>(0);          // 金额

    const handleDateChange = (value: any, dateString: string) => {
        setBillDate(new Date(dateString));
    }
    const handleCategoryChange = (categoryID: string) => {
        setBillCategory(categoryID);
    }
    const handleAmountChange = (amount: any) => {
        setBillAmount(amount as number);
    }
    const handleBillAdd = () => {
        if (billTime.getTime() === (new Date(0).getTime())) {
            message.warn('time must be selected ~');
            return;
        }
        if (billCategory === '') {
            message.warn('category must be selected ~');
            return;
        }
        if (billAmount === 0) {
            message.warn('amount cant be ￥0.00 ~');
            return;
        }

        const c = props.dataCategory.find(v => v.id === billCategory);  // 为了该 category 是什么 type
        c && props.onClick({
            time: billTime,
            type: c.type,
            category: billCategory,
            amount: billAmount
        });
        message.success('success ~');
    }

    return (
        <div className="bill-add">
            <DatePicker className="bill-add-part" showTime onChange={handleDateChange} />
            <Select className="bill-add-part" placeholder='Select category' onChange={handleCategoryChange}>
                {props.dataCategory.map((v, i) => {
                    return <Select.Option key={i} value={v.id}>{v.name}</Select.Option>
                })}
            </Select>
            <InputNumber
                className="bill-add-part"
                precision={2}
                defaultValue={0}
                formatter={value => `￥${value}`.replace(/\B(?=(\d{3})+(?!\d))/g, ',')}
                parser={value => value ? value.replace(/￥\s?|(,*)/g, '') : 0}
                onChange={handleAmountChange}
            />
            <Button className="bill-add-part" type='primary' onClick={handleBillAdd}>添加账单</Button>
        </div>
    );
};

export default BillAdd;
