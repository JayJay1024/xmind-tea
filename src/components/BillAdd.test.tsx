import React from 'react';
import { RenderResult, render, act, screen } from '@testing-library/react';

import BillAdd from './BillAdd';

describe('<BillAdd />', () => {
    let renderResult: RenderResult;

    beforeEach(() => {
        act(() => {
            renderResult = render(<BillAdd dataCategory={[]} onClick={() => {}} />);
        });
    });

    test('there should be a date picker', () => {
        expect(screen.getByPlaceholderText('Select date')).toBeInTheDocument();
    });
    test('there should be a catory selector', () => {
        expect(screen.getByRole('combobox')).toBeInTheDocument();
    });

    test('there should be an amount input with default value of ￥0.00', () => {
        expect(screen.getByRole('spinbutton')).toBeInTheDocument();
        expect(screen.getByRole('spinbutton')).toHaveValue('￥0.00');
    });

    test('there should be a button to add bill', () => {
        expect(screen.getByText('添加账单')).toBeInTheDocument();
    });

    test('should match snapshot', () => {
        expect(renderResult.asFragment()).toMatchSnapshot();
    });
});

