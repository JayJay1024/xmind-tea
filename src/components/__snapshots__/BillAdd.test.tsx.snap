// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BillAdd /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="bill-add"
  >
    <div
      class="ant-picker bill-add-part"
    >
      <div
        class="ant-picker-input"
      >
        <input
          placeholder="Select date"
          readonly=""
          size="21"
          title=""
          value=""
        />
        <span
          class="ant-picker-suffix"
        >
          <span
            aria-label="calendar"
            class="anticon anticon-calendar"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="calendar"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="ant-select bill-add-part ant-select-single ant-select-show-arrow"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-placeholder"
        >
          Select category
        </span>
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
    <div
      class="ant-input-number bill-add-part"
    >
      <div
        class="ant-input-number-handler-wrap"
      >
        <span
          aria-label="Increase Value"
          class="ant-input-number-handler ant-input-number-handler-up"
          role="button"
          unselectable="unselectable"
        >
          <span
            aria-label="up"
            class="anticon anticon-up ant-input-number-handler-up-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="up"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
              />
            </svg>
          </span>
        </span>
        <span
          aria-label="Decrease Value"
          class="ant-input-number-handler ant-input-number-handler-down"
          role="button"
          unselectable="unselectable"
        >
          <span
            aria-label="down"
            class="anticon anticon-down ant-input-number-handler-down-inner"
            role="img"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </span>
      </div>
      <div
        class="ant-input-number-input-wrap"
      >
        <input
          aria-valuemin="-9007199254740991"
          aria-valuenow="0"
          autocomplete="off"
          class="ant-input-number-input"
          min="-9007199254740991"
          role="spinbutton"
          step="1"
          value="￥0.00"
        />
      </div>
    </div>
    <button
      class="ant-btn bill-add-part ant-btn-primary"
      type="button"
    >
      <span>
        添加账单
      </span>
    </button>
  </div>
</DocumentFragment>
`;
